clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -1839,6 +1839,41 @@ static void handleRestrictAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
                  RestrictAttr(S.Context, AL, DeallocE, DeallocPtrIdx));
 }
 
+static bool isSpanLikeType(const QualType &Ty) {
+  // Check that the type is a plain record with one field being a pointer
+  // type and the other field being an integer. This matches the common
+  // implementation of std::span or sized_allocation_t in P0901R11.
+  // Note that there may also be numerous cases of pointer+integer structures
+  // not actually exhibiting a span-like semantics, so sometimes
+  // this heuristic expectedly leads to false positive results.
+  const RecordDecl *RD = Ty->getAsRecordDecl();
+  if (!RD || RD->isUnion())
+    return false;
+  const RecordDecl *Def = RD->getDefinition();
+  if (!Def)
+    return false; // This is an incomplete type.
+  auto FieldsBegin = Def->field_begin();
+  if (std::distance(FieldsBegin, Def->field_end()) != 2)
+    return false;
+  const QualType FirstFieldType = FieldsBegin->getType();
+  const QualType SecondFieldType = std::next(FieldsBegin)->getType();
+  // Verify two possible orderings.
+  return (FirstFieldType->isAnyPointerType() &&
+          SecondFieldType->isIntegerType()) ||
```

Comment:
I think we should be stricter on integer type.  Also, are we sure we mean `isAnyPointerType`?  This includes some odd ObjC types that I'm not sure we mean.  We should also do some work here (perhaps just writing tests) to ensure that we aren't allowing function member/data member pointers.

I might suggest we also do some work to disallow function pointers unless there is a REALLY good reason not to.

Reply:
I would do something more restrictive.  Probably an `isSameType` as `Context::PtrDiffT`.

Additionally, I don't see anything in here that prevents these from being bitfields!  That seems like it would be wrong too.